src/main/java/com/github/princesslana/eriscasper/data/resource/AuditLogChange.java

```diff
@@ -0,0 +1,163 @@
+package com.github.princesslana.eriscasper.data.resource;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.github.princesslana.eriscasper.data.Data;
+import com.github.princesslana.eriscasper.data.DataException;
+import com.github.princesslana.eriscasper.data.Snowflake;
+import com.google.common.collect.ImmutableList;
+import java.util.Optional;
+
+@SuppressWarnings("unused")
+public class AuditLogChange<T> {
+
+  private Class<T> valueClass;
+  private T newValue;
+  private T oldValue;
+  private String key;
+
+  private AuditLogChange(Class<T> valueClass, T newValue, T oldValue, String key) {
+    this.valueClass = valueClass;
+    this.newValue = newValue;
+    this.oldValue = oldValue;
+    this.key = key;
+  }
+
+  public Class<T> getValueClass() {
+    return valueClass;
+  }
+
+  public boolean isArray() {
+    return valueClass == null;
+  }
+
+  public Optional<T> getNewValue() {
+    return Optional.ofNullable(newValue);
+  }
+
+  public Optional<T> getOldValue() {
+    return Optional.ofNullable(oldValue);
+  }
+
+  public String getKey() {
+    return key;
+  }
+
+  public static AuditLogChange<?> parseJson(JsonNode node) {
```

Comment:
Exposing with wildcard generics is a code smell. There's no way anyone can know what `T` is, so why have it as a generic parameter at all?

Reply:
Yeah just removed all generics.